ksql-serde/src/main/java/io/confluent/ksql/serde/json/KsqlJsonDeserializer.java

```diff
@@ -115,66 +137,94 @@ private static Object enforceFieldType(final JsonValueContext context) {
   }
 
   private static List<?> enforceElementTypeForArray(final JsonValueContext context) {
-    if (!(context.val instanceof List)) {
+    if (!(context.val instanceof ArrayNode)) {
       throw invalidConversionException(context.val, context.schema);
     }
 
-    final List<?> list = (List<?>) context.val;
+    final ArrayNode list = (ArrayNode) context.val;
     final List<Object> array = new ArrayList<>(list.size());
-    for (final Object item : list) {
+    for (final JsonNode item : list) {
       array.add(enforceFieldType(context.deserializer, context.schema.valueSchema(), item));
     }
     return array;
   }
 
   private static Map<String, Object> enforceKeyAndValueTypeForMap(final JsonValueContext context) {
-    if (!(context.val instanceof Map)) {
+    if (!(context.val instanceof ObjectNode)) {
       throw invalidConversionException(context.val, context.schema);
     }
 
-    final Map<?, ?> map = (Map<?, ?>) context.val;
+    final ObjectNode map = (ObjectNode) context.val;
     final Map<String, Object> ksqlMap = new HashMap<>(map.size());
-    for (final Map.Entry<?, ?> e : map.entrySet()) {
+    for (Iterator<Entry<String, JsonNode>> it = map.fields(); it.hasNext(); ) {
+      final Entry<String, JsonNode> e = it.next();
       ksqlMap.put(
           enforceFieldType(
-              context.deserializer, Schema.OPTIONAL_STRING_SCHEMA, e.getKey()).toString(),
+              context.deserializer,
+              Schema.OPTIONAL_STRING_SCHEMA,
+              new TextNode(e.getKey()))
+              .toString(),
           enforceFieldType(
               context.deserializer, context.schema.valueSchema(), e.getValue())
       );
     }
     return ksqlMap;
   }
 
-  @SuppressWarnings("unchecked")
   private static Struct enforceFieldTypesForStruct(final JsonValueContext context) {
-    if (!(context.val instanceof Map)) {
+    if (!(context.val instanceof ObjectNode)) {
       throw invalidConversionException(context.val, context.schema);
     }
 
     final Struct columnStruct = new Struct(context.schema);
-    final Map<String, ?> jsonFields = (Map<String, ?>) context.val;
-
-    final Map<String, ?> upperCasedFields = upperCaseKeys(jsonFields);
+    final ObjectNode jsonFields = (ObjectNode) context.val;
+    final Map<String, JsonNode> upperCasedFields = upperCaseKeys(jsonFields);
 
     for (Field ksqlField : context.schema.fields()) {
       // the "case insensitive" strategy leverages that all KSQL fields are internally
       // case sensitive - if they were specified without quotes, then they are upper-cased
       // during parsing. any ksql fields that are case insensitive, therefore, will be matched
       // in this case insensitive field map without modification but the quoted fields will not
       // (unless they were all uppercase to start off with, which is expected to match)
-      final Object fieldValue = ObjectUtils.defaultIfNull(
+      final JsonNode fieldValue = ObjectUtils.defaultIfNull(
```

Comment:
side note: 

Shame we pay the price of the lookup into `upperCasedFields` even if `jsonFields.get(ksqlField.name())` returns non-null.

Reply:
+1 good catch